OpenSim/Moco/MocoFrameOrientationConstraint.cpp

```diff
@@ -0,0 +1,127 @@
+/* -------------------------------------------------------------------------- *
+ * OpenSim Moco: MocoFrameOrientationConstraint.cpp                           *
+ * -------------------------------------------------------------------------- *
+ * Copyright (c) 2022 Stanford University and the Authors                     *
+ *                                                                            *
+ * Author(s): Size Zheng                                                      *
+ *                                                                            *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may    *
+ * not use this file except in compliance with the License. You may obtain a  *
+ * copy of the License at http://www.apache.org/licenses/LICENSE-2.0          *
+ *                                                                            *
+ * Unless required by applicable law or agreed to in writing, software        *
+ * distributed under the License is distributed on an "AS IS" BASIS,          *
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   *
+ * See the License for the specific language governing permissions and        *
+ * limitations under the License.                                             *
+ * -------------------------------------------------------------------------- */
+
+#include "MocoFrameOrientationConstraint.h"
+
+using namespace OpenSim;
+
+
+MocoFrameOrientationConstraintPair::MocoFrameOrientationConstraintPair() {
+    constructProperties();
+}
+
+MocoFrameOrientationConstraintPair::MocoFrameOrientationConstraintPair(
+        std::string frame1Path, std::string frame2Path, double minimum_angle, 
+        double maximum_angle) {
+    constructProperties();
+    set_frame1_path(frame1Path);
+    set_frame2_path(frame2Path);
+    set_minimum_angle(minimum_angle);
+    set_maximum_angle(maximum_angle);
+}
+
+void MocoFrameOrientationConstraintPair::constructProperties() {
+    constructProperty_frame1_path("");
+    constructProperty_frame2_path("");
+    constructProperty_minimum_angle(-SimTK::Infinity);
+    constructProperty_maximum_angle(SimTK::Infinity);
+}
+
+//=============================================================================
+//  MocoFrameOrientationConstraint
+//=============================================================================
+
+MocoFrameOrientationConstraint::MocoFrameOrientationConstraint() {
+    constructProperties();
+}
+void MocoFrameOrientationConstraint::initializeOnModelImpl(const Model& model, 
+        const MocoProblemInfo&) const {
+
+    int nFramePairs = getProperty_frame_pairs().size();
+    MocoConstraintInfo info;
+    std::vector<MocoBounds> bounds;
+
+    for (int i = 0; i < nFramePairs; ++i) {

```

Comment:
You could use the concise for loop style here: `for (const auto& frame_pair : m_frame_pairs)`.